libphal/phal_pdbg.C

```diff
@@ -300,24 +301,134 @@ bool isSbeVitalAttnActive(struct pdbg_target *proc)
-  if (pib_read(pibTarget, l_coreScratchRegAddr, &l_coreScratchRegData))
-  {
-    // If unable to read the register, by default consider the control is in hostboot
-    log(level::ERROR, "scom read error: 0x%X", l_coreScratchRegAddr );
-    return false;
-  }
-
-  // If the register reads zero, return control moved to host.
-  return (l_coreScratchRegData == 0);
+	// Read the scratch register to find if the control has moved to phyp
+	// (Host)
+	auto pibTarget = pdbg_target_from_path(nullptr, "/proc0/pib");
+
+	uint64_t l_coreScratchRegData = 0xFFFFFFFFFFFFFFFFull;
+	// HB changes the below core scratch reg as one of the last instructions
+	// that is run, so if that is zero then we're in phyp
+	uint64_t l_coreScratchRegAddr = 0x4602F489;
+
+	// Is there any error in reading the scom address, consider control is
+	// in hostboot
+	if (pib_read(pibTarget, l_coreScratchRegAddr, &l_coreScratchRegData)) {
+		// If unable to read the register, by default consider the
+		// control is in hostboot
+		log(level::ERROR, "scom read error: 0x%X",
+		    l_coreScratchRegAddr);
+		return false;
+	}
+
+	// If the register reads zero, return control moved to host.
+	return (l_coreScratchRegData == 0);
+}
+
+using LocationCode = std::string;
+using FRUType = uint8_t;
+constexpr uint8_t EXP_LOCATIN_CODE_MIN_LENGTH = 17;
+constexpr uint8_t UNEXP_LOCATIN_CODE_MIN_LENGTH = 4;
+constexpr int continueTgtTraversal = 0;
+constexpr int reqAttrFound = 1;
+constexpr int reqAttrNotFound = 2;
+LocationCode g_locCode;
+
+std::optional<LocationCode> getUnexpandedLocCode(const std::string &locCode)
+{
+
+	// "-" should be there to seggregate all (FC, Node number and SE values)
+	// together.
+	// Note: Each (FC, Node number and SE) value can be seggregate by "."
+	// but, cec device tree just have unexpand format so, just skipping
+	// still first occurrence "-" and replacing with "fcs".
+	auto endPosOfFcs = locCode.find('-', EXP_LOCATIN_CODE_MIN_LENGTH);
+	if (endPosOfFcs == std::string::npos) {
+		log(level::ERROR,
+		    "Given location code %s is not valid i.e could not find "
+		    "dash",
+		    locCode);
+		return std::nullopt;
+	}
+
+	std::string unExpandedLocCode("Ufcs");
+
+	if (locCode.length() > EXP_LOCATIN_CODE_MIN_LENGTH) {
+		unExpandedLocCode.append(
+		    locCode.substr(endPosOfFcs, std::string::npos));
+	}
+
+	return unExpandedLocCode;
+}
+
+int pdbgCallbackToGetLocCode(struct pdbg_target *target, void *fruType)
+{
+	ATTR_LOCATION_CODE_Type locCode;
+	std::vector<uint8_t> frus{ENUM_ATTR_TYPE_DIMM, ENUM_ATTR_TYPE_PROC,
+				  ENUM_ATTR_TYPE_TPM, ENUM_ATTR_TYPE_BMC};
+	if (DT_GET_PROP(ATTR_LOCATION_CODE, target, locCode))
+		return continueTgtTraversal;
+	else {
+		if (locCode == g_locCode) {
+			if (!DT_GET_PROP(ATTR_TYPE, target,
+					 *static_cast<int *>(fruType)) &&
+			    std::find(frus.begin(), frus.end(),
+				      *static_cast<int *>(fruType)) !=
+				frus.end())
+				return reqAttrFound;
+			return reqAttrNotFound;
```

Comment:
Actually, `reqAttrNotFound` is not a valid error code here because there is a chance that ATTR_TYPE might not be readable or not found. ATTR_TYPE will be available in all targets since it is a base attribute for the target. Additionally, the ATTR_TYPE value might not match the predefined FRUs list. Therefore, we need to be able to catch a bug if the user passes a valid FRU location code that is not in the FRUs list you declared. We should have a proper error code for each case to accurately identify the issue.